src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx

```diff
@@ -91,21 +79,105 @@ const DiffExprCompute = (props) => {
       comparisonGroup[selectedComparison]['basis'] = `sample/${samples[0].key}`
     }
 
+    setSampleKeys(samples.map(sample => sample.key));
+
   }, [hierarchy, properties]);
 
-  const validateForm = () => {
+  const cellIdToSampleMap = useMemo(() => {
+    const mapping = [];
+    sampleKeys.forEach((key, idx) => {
+      const cellIds = properties[key].cellIds;
+      cellIds.forEach(cellId => mapping[cellId] = idx);
+    });
 
-    if (!comparisonGroup[selectedComparison]?.cellSet) {
-      setIsFormValid(false);
-      return;
+    return mapping;
+  }, [numSamples]);
+
+  const canRunDiffExpr = useCallback(() => {
+
+    if (selectedComparison === ComparisonType.WITHIN) return true;
+
+    if (!basis || !cellSet || !compareWith || !cellIdToSampleMap.length > 0) { return false; }
+
+    const basisCellSetKey = getCellSetName(basis);
+
+    // Group 1 is from cellSet
+    // Group 2 is from compareWith
+    const group1CellSetKey = getCellSetName(cellSet);
+    const group2CellSetKey = getCellSetName(compareWith);
+
+    let basisCellSet = [];
+    if (basisCellSetKey === 'all') {
+      const allCellIds = sampleKeys.reduce((acc, key) => {
+        const cellIds = properties[key].cellIds;
+        return acc.concat(Array.from(cellIds));
+      }, []);
+      basisCellSet = new Set(allCellIds);
+    } else {
+      basisCellSet = properties[basisCellSetKey].cellIds;
+    }
+
+    const group1CellIds = Array.from(properties[group1CellSetKey].cellIds);
+
+    let group2CellIds = [];
+    if (['rest', 'background'].includes(group2CellSetKey)) {
+      const parentGroup = getRootKey(cellSet);
+
+      const otherGroupKeys = hierarchy.find(obj => obj.key === parentGroup)
+        .children.filter(child => child.key !== group1CellSetKey);
+
+      group2CellIds = otherGroupKeys.reduce((acc, child) => {
+        return acc.concat(Array.from(properties[child.key].cellIds));
+      }, []);
+    } else {
+      group2CellIds = Array.from(properties[group2CellSetKey].cellIds);
     }
 
-    if (!comparisonGroup[selectedComparison]?.compareWith) {
+    // Intersect the basis cell set with each group cell set
+    const filteredGroup1CellIds = group1CellIds.filter(cellId => basisCellSet.has(cellId));
+    const filteredGroup2CellIds = group2CellIds.filter(cellId => basisCellSet.has(cellId));
+
+    // Prepare an array of length sampleIds to hold tally of cells for each sapmple in each group
+    const cellsInGroup1Samples = new Array(numSamples).fill(0);
```

Comment:
Can you not have a loop with those groups so you do all the operations only once in the loop not once for each group manually?

Reply:
If you think thats a good idea, doesnt matter too much at the end

Reply:
Yea, it can be done like that, storing the groups in an object, and then using it. I just thought it's gonna be clearer this way, because there are only 2 groups and we can see what we're doing.